public/app/components/visualizations/HorizontalStackedBarChart.jsx

```diff
@@ -0,0 +1,166 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import * as d3 from 'd3';
+
+
+class HorizontalStackedBarChart extends React.Component {
+
+  constructor(props) {
+    super(props);
+
+    this.renderChart = this.renderChart.bind(this);
+
+    this.stack = d3.stack();
+    this.color = d3.scaleOrdinal(props.colors);
+
+    const container = {
+      width: 500,
+      height: 500,
+    };
+
+    const margin = {
+      top: 20,
+      right: 20,
+      bottom: 50,
+      left: 60
+    };
+
+    this.size = {
+      width: (container.width - margin.left) - margin.right,
+      height: (container.height - margin.top) - margin.bottom,
+      margin,
+      container,
+    };
+  }
+
+  renderChart() {
+    const { width, height, margin, container } = this.size;
+
+    const x = d3.scaleLinear().range([0, width]);
+    const y = d3.scaleBand()
+      .domain(d3.extent(this.props.data, d => d.y))
+      .range([height, 0])
+      .paddingInner(0.2);
+
+    this.gChart.attr('transform', `translate(${margin.left},${margin.top})`);
+
+    const keys = [...(new Set(this.props.data.map(d => d.z)))];
+
+    this.color.domain(keys);
+    this.stack.keys(keys);
+
+    let data = this.props.data.reduce((acc, row) => {
+        acc[row.y] = { ...(acc[row.y] || {}), ...{[row.z]: row.x} };
+        return acc;
+      }, {});
```

Comment:
I think this could be slightly cleaner as a scrap scoped forEach. In a perfect world, an accumulator would be immutable.

Reply:
Something like this?

```js
let data = (() => {
  const acc = {};
  this.props.data.forEach(row => acc[row.y] = {...(acc[row.y] || {}), ...{[row.z]: row.x}});
  return acc;
})();
```